mlir/lib/IR/AffineExpr.cpp

```diff
@@ -237,6 +239,144 @@ unsigned AffineDimExpr::getPosition() const {
   return static_cast<ImplType *>(expr)->position;
 }
 
+/// Returns the divide type which are as follow floorrdiv = 1, ceildiv = 2,
+/// and otherwise 0.
+static int getDivideType(AffineExprKind exprKind) {
+  if (exprKind == AffineExprKind::FloorDiv)
+    return 1;
+  if (exprKind == AffineExprKind::CeilDiv)
+    return 2;
+  return 0;
+}
+
+/// Returns true if the expression is divisible by the given symbol.
+static bool isDivisibleBySymbol(AffineExpr expr, unsigned symbolPos,
+                                int divideType) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+  case AffineExprKind::DimId:
+    return false;
+  case AffineExprKind::SymbolId:
+    return (expr.cast<AffineSymbolExpr>().getPosition() == symbolPos);
+  // Checking divisibility by given symbol for both operands.
+  case AffineExprKind::Add: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, divideType) &&
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, divideType);
+  }
+  // Operands of modulo if are floordiv are complicated to be simplify.
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, 0) &&
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, 0);
+  }
+  // Checking if any one of the operand divisible by given symbol.
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, divideType) ||
+           isDivisibleBySymbol(binaryExpr.getRHS(), symbolPos, divideType);
+  }
+  // Floordiv and ceildiv are not divisible by symbol because of already
+  // existing division.
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::CeilDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    int type = getDivideType(expr.getKind());
+    if (divideType != type)
+      return false;
+    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, type);
+  }
+  }
+  llvm_unreachable("Unknown AffineExpr");
+}
+
+/// Divides a given expression by a given symbol. It considers all divisibility
```

Comment:
This is not just dividing, but dividing while simplifying. Make doc comment accurate.